### PR TITLE
cache: fix flaky resync test

### LINKED ISSUE #25953: Unit test flake: TestReflectorResync exactly 2 iterations were expected, got: 3
```
-- FAIL: TestReflectorResync (0.01s)
	reflector_test.go:380: exactly 2 iterations were expected, got: 3
```
https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/25904/kubernetes-pull-test-unit-integration/26834/

### PR DESCRIPTION
**Pull request by xiang90**
close(stopch) will not stop ListAndWatch in a sync manner since chan recv is async. Thus is test suffers from both logic race and data race.

This patch changes the exit notification to an error. 

Fix https://github.com/kubernetes/kubernetes/issues/25953

/cc @wojtek-t @ncdc

### CODE REVIEW

**Comment by xiang90** [comment]
@k8s-bot test this github flake: #25919

**Comment by xiang90** [comment]
@k8s-bot unit test this issue: #25967

**Comment by wojtek-t** [comment]
LGTM - thanks @xiang90 !

**Comment by piosz** [comment]
Merging manually to unblock submit queue.

File: pkg/client/cache/reflector_test.go [hunk 1]
```diff
@@ -354,11 +355,12 @@ func TestReflectorListAndWatchWithErrors(t *testing.T) {
 func TestReflectorResync(t *testing.T) {
 	iteration := 0
 	stopCh := make(chan struct{})
+	rerr := errors.New("expected resync reached")
 	s := &FakeCustomStore{
 		ResyncFunc: func() error {
 			iteration++
 			if iteration == 2 {
-				close(stopCh)
+				return rerr
```
Comment (kargakis) [inline pkg/client/cache/reflector_test.go]: It seems that this error is never returned from [ListAndWatch](https://github.com/kubernetes/kubernetes/blob/8f104a7b0f2298caaf5f8a4ab901855646172886/pkg/client/cache/reflector.go#L336).

Comment (kargakis) [inline pkg/client/cache/reflector_test.go]: https://github.com/kubernetes/kubernetes/issues/26065

**Pull request merged**